File: GameLibrary/src/cc/lib/swing/AWTMonopoly.java
package cc.lib.swing;

import java.awt.BorderLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.File;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.swing.JLabel;
import javax.swing.JSpinner;
import javax.swing.SpinnerNumberModel;
import javax.swing.event.ChangeEvent;

import cc.lib.game.AGraphics;
import cc.lib.game.Utils;
import cc.lib.logger.Logger;
import cc.lib.logger.LoggerFactory;
import cc.lib.monopoly.Card;
import cc.lib.monopoly.MoveType;
import cc.lib.monopoly.Piece;
import cc.lib.monopoly.Player;
import cc.lib.monopoly.PlayerUser;
import cc.lib.monopoly.Rules;
import cc.lib.monopoly.Square;
import cc.lib.monopoly.Trade;
import cc.lib.monopoly.UIMonopoly;
import cc.lib.utils.FileUtils;
import cc.lib.utils.Reflector;

public class AWTMonopoly extends AWTComponent {
    private final static Logger log = LoggerFactory.getLogger(AWTMonopoly.class);

    public static void main(String[] args) {
        AGraphics.DEBUG_ENABLED = true;
        Utils.setDebugEnabled();
        new AWTMonopoly();
    }

    final File SAVE_FILE;
    final File RULES_FILE;

    final AWTFrame frame;
    final UIMonopoly monopoly = new UIMonopoly() {
        @Override
        public void repaint() {
            AWTMonopoly.this.repaint();
        }

        @Override
        public int getImageId(Piece p) {
            return pieceMap.get(p);
        }

        @Override
        public int getBoardImageId() {
            return ids[Images.monopoly_board.ordinal()];
        }

        @Override
        protected MoveType showChooseMoveMenu(final Player player, final List<MoveType> moves) {
            final MoveType [] result = new MoveType[1];
            final String [] strings = new String[moves.size()];
            int index = 0;
            for (MoveType mt : moves) {
                switch (mt) {
                    case PURCHASE_UNBOUGHT: {
                        Square sq = getPurchasePropertySquare();
                        strings[index++] = "Purchase " + Utils.toPrettyString(sq.name()) + " $" + sq.getPrice();
                        break;
                    }
                    case PAY_BOND: {
                        strings[index++] = String.format("%s $%d", Utils.toPrettyString(mt.name()), player.getJailBond());
                        break;
                    }
                    default:
                        strings[index++] = Utils.toPrettyString(mt.name());
                        break;
                }
            }
            monopoly.trySaveToFile(SAVE_FILE);
            frame.showListChooserDialog(new AWTFrame.OnListItemChoosen() {
                @Override
                public void itemChoose(int index) {
                    result[0] = moves.get(index);
                    switch (result[0]) {
                        case PURCHASE_UNBOUGHT:{
                            showPropertyPopup("Purchase?", "Buy",
                                    getPurchasePropertySquare(), () -> {
                                        synchronized (monopoly) {
                                            monopoly.notify();
                                        }
                                    }, () -> {
                                        result[0] = null;
                                        synchronized (monopoly) {
                                            monopoly.notify();
                                        }

                                    }
                            );
                            break;
                        }
                        case PURCHASE: {
                            showPropertyPopup("Purchase?", "Buy", 
                                    player.getSquare(), () -> {
                                        synchronized (monopoly) {
                                            monopoly.notify();
                                        }
                                    }, () -> {
                                        result[0] = null;
                                        synchronized (monopoly) {
                                            monopoly.notify();
                                        }

                                    }
                            );
                            break;
                        }
                        default:
                            synchronized (monopoly) {
                                monopoly.notify();
                            }
                    }
                }

                @Override
                public void cancelled() {
                    if (monopoly.canCancel())
                        monopoly.cancel();
                    else
                        monopoly.stopGameThread();
                    synchronized (monopoly) {
                        monopoly.notify();
                    }
                }
            }, "Choose Move " + player.getPiece(), strings);
            Utils.waitNoThrow(monopoly, -1);
            return result[0];
        }

        @Override
        protected Card showChooseCardMenu(final Player player, final List<Card> cards, final Player.CardChoiceType type) {
            final Card [] result = new Card[1];
            final String [] items = new String[cards.size()];
            int index = 0;
            for (Card c : cards) {
                switch (type) {

                    case CHOOSE_CARD_TO_MORTGAGE:
                        items[index++] = Utils.toPrettyString(c.getProperty().name()) + " Mortgage $" + c.getProperty().getMortgageValue(c.getHouses());
                        break;
                    case CHOOSE_CARD_TO_UNMORTGAGE:
                        items[index++] = Utils.toPrettyString(c.getProperty().name()) + " Buyback $" + c.getProperty().getMortgageBuybackPrice();
                        break;
                    case CHOOSE_CARD_FOR_NEW_UNIT:
                        items[index++] = Utils.toPrettyString(c.getProperty().name()) + " Unit $" + c.getProperty().getUnitPrice();
                        break;
                }
            }
            frame.showListChooserDialog(new AWTFrame.OnListItemChoosen() {
                @Override
                public void itemChoose(int index) {
                    result[0] = cards.get(index);
                    synchronized (monopoly) {
                        monopoly.notify();
                    }
                }

                @Override
                public void cancelled() {
                    if (monopoly.canCancel())
                        monopoly.cancel();
                    else
                        monopoly.stopGameThread();
                    synchronized (monopoly) {
                        monopoly.notify();
                    }
                }
            }, player.getPiece() +  " " + Utils.toPrettyString(type.name()), items);
            Utils.waitNoThrow(monopoly, -1);
            return result[0];
        }

        @Override
        protected Trade showChooseTradeMenu(Player player, List<Trade> trades) {
            final Trade [] result = new Trade[1];
            frame.showListChooserDialog(new AWTFrame.OnListItemChoosen() {
                @Override
                public void itemChoose(int index) {
                    final Trade t = trades.get(index);
                    final Player trader = t.getTrader();
                    String title = "Buy " + Utils.toPrettyString(t.getCard().getProperty().name()) + " from " + Utils.toPrettyString(trader.getPiece().name());
                    showPropertyPopup(title, "Buy for $" +t.getPrice(), t.getCard().getProperty(), () -> {
                        result[0] = t;
                        synchronized (monopoly) {
                            monopoly.notify();
                        }
                    }, ()->{
                        synchronized (monopoly) {
                            monopoly.notify();
                        }
                    });
                }

                @Override
                public void cancelled() {
                    if (monopoly.canCancel())
                        monopoly.cancel();
                    else
                        monopoly.stopGameThread();
                    synchronized (monopoly) {
                        monopoly.notify();
                    }
                }
            }, "Choose Trade" + player.getPiece(), Utils.toStringArray(trades, true));
            Utils.waitNoThrow(monopoly, -1);
            return result[0];
        }

        @Override
        protected boolean showMarkSellableMenu(PlayerUser user, List<Card> sellable) {
            final AWTFrame popup = new AWTFrame();
            final ActionListener listener = (ActionEvent e) -> {
                synchronized (monopoly) {
                    monopoly.notify();
                }
                popup.closePopup();
            };
            AWTPanel content = new AWTPanel(new BorderLayout());
            AWTPanel list = new AWTPanel(0, 2);
            for (final Card card : sellable) {
                int price = user.getSellableCardCost(card);
                list.add(new JLabel(card.getProperty().name()));
                final JSpinner spinner = new JSpinner(new SpinnerNumberModel(price, 0, 2000, 50));
                spinner.addChangeListener((ChangeEvent e) -> {
                    user.setSellableCard(card, (Integer)spinner.getValue());
                });
                list.add(spinner);
            }
            content.add(new AWTLabel("Mark Sellable " + user.getPiece().name(), 1, 20, true), BorderLayout.NORTH);
            content.add(list, BorderLayout.CENTER);
            content.add(new AWTButton("DONE", listener), BorderLayout.SOUTH);
            popup.setContentPane(content);
            popup.showAsPopup(frame);
            Utils.waitNoThrow(monopoly, -1);
            return true;
        }
    };

    public AWTFrame showPropertyPopup(String title, String yesLabel, final Square sq, final Runnable onBuyRunnable, final Runnable onDontBuyRunnble) {
        final AWTFrame popup = new AWTFrame();
        AWTPanel content = new AWTPanel(new BorderLayout());
        content.add(new AWTLabel(title, 1, 20, true), BorderLayout.NORTH);
        AWTComponent sqPanel = new AWTComponent() {
            @Override
            protected void paint(AWTGraphics g, int mouseX, int mouseY) {
                monopoly.drawPropertyCard(g, sq, g.getViewportWidth(), g.getViewportHeight());
            }
        };
        sqPanel.setPreferredSize(monopoly.DIM/3, monopoly.DIM/4*2);
        content.add(sqPanel);
        content.add(new AWTPanel(1, 2, new AWTButton(yesLabel) {
            @Override
            protected void onAction() {
                popup.closePopup();
                onBuyRunnable.run();
            }
        }, new AWTButton("Dont Buy") {
            @Override
            protected void onAction() {
                popup.closePopup();
                onDontBuyRunnble.run();
            }
        }), BorderLayout.SOUTH);
        popup.setContentPane(content);
        popup.showAsPopup(frame);
        return popup;
    }

    AWTMonopoly() {
        setMouseEnabled(true);
        setPadding(5);
        frame = new AWTFrame("Monopoly") {
            /*
            @Override
            protected void onWindowClosing() {
                if (monopoly.isGameRunning())
                    monopoly.trySaveToFile(SAVE_FILE);
            }*/

            @Override
            protected void onMenuItemSelected(String menu, String subMenu) {
                switch (menu) {
                    case "File":
                        switch (subMenu) {
                            case "New Game":
                                frame.showListChooserDialog(new OnListItemChoosen() {
                                    @Override
                                    public void itemChoose(final int item) {
                                        frame.showListChooserDialog(new OnListItemChoosen() {
                                            @Override
                                            public void itemChoose(int pc) {
                                                monopoly.stopGameThread();
                                                monopoly.initPlayers(item + 2, Piece.values()[pc]);
                                                monopoly.newGame();
                                                monopoly.startGameThread();
                                            }

                                            @Override
                                            public void cancelled() {}
                                        }, "Choose Piece", Utils.toStringArray(Piece.values(), true));

                                    }
                                    @Override
                                    public void cancelled() {}
                                }, "How Many Players?", "2", "3", "4");
                                break;
                            case "Resume":
                                if (monopoly.tryLoadFromFile(SAVE_FILE)) {
                                    monopoly.startGameThread();
                                }
                                break;
                            case "Rules": {
                                final Rules rules = monopoly.getRules();
                                final AWTFrame popup = new AWTFrame();
                                AWTPanel content= new AWTPanel(new BorderLayout());
                                final AWTNumberPicker npStart = new AWTNumberPicker.Builder().setLabel("Start $").setMin(500).setMax(1500).setStep(100).setValue(rules.startMoney).build(null);
                                final AWTNumberPicker npWin   = new AWTNumberPicker.Builder().setLabel("Win $").setMin(2000).setMax(5000).setStep(500).setValue(rules.valueToWin).build(null);
                                final AWTNumberPicker npTaxScale = new AWTNumberPicker.Builder().setLabel("Tax Scale %").setMin(0).setMax(200).setStep(50).setValue(Math.round(100f * rules.taxScale)).build(null);
                                final AWTToggleButton jailBump = new AWTToggleButton("Jail Bump", rules.jailBumpEnabled);
                                final AWTToggleButton jailMulti = new AWTToggleButton("Jail Multiplier", rules.jailMultiplier);
                                content.addTop(new AWTLabel("RULES", 1, 20, true));
                                AWTPanel buttons = new AWTPanel(0, 1);
                                AWTPanel pickers = new AWTPanel(1, 0);
                                pickers.add(npStart);
                                pickers.add(npWin);
                                pickers.add(npTaxScale);
                                content.add(pickers);
                                content.addRight(buttons);
                                buttons.add(jailBump);
                                buttons.add(jailMulti);
                                buttons.add(new AWTButton("Apply") {
                                    @Override
                                    protected void onAction() {
                                        rules.valueToWin = npWin.getValue();
                                        rules.taxScale = 0.01f * npTaxScale.getValue();
                                        rules.startMoney = npStart.getValue();
                                        rules.jailBumpEnabled = jailBump.isSelected();
                                        rules.jailMultiplier = jailMulti.isSelected();
                                        popup.closePopup();
                                        try {
                                            Reflector.serializeToFile(rules, RULES_FILE);
                                        } catch (Exception e) {
                                            frame.showMessageDialog("ERROR", "Error save rules to file " + RULES_FILE + "\n" + e.getClass().getSimpleName() + " " + e.getMessage());
                                        }
                                    }
                                });
                                content.addBottom(new AWTButton("Cancel") {
                                    @Override
                                    protected void onAction() {
                                        popup.closePopup();
                                    }
                                });
                                popup.setContentPane(content);
                                popup.showAsPopup(frame);
                                break;
                            }
                        }
                }
            }
        };

        frame.addMenuBarMenu("File", "New Game", "Resume", "Rules");
        frame.add(this);

        File settings = FileUtils.getOrCreateSettingsDirectory(getClass());
        SAVE_FILE = new File(settings, "game.save");
        RULES_FILE = new File(settings, "rules.save");
        if (!frame.loadFromFile(new File(settings, "monopoly.properties")))
            frame.centerToScreen(800, 600);
        if (RULES_FILE.isFile()) {
            try {
                Rules rules = Reflector.deserializeFromFile(RULES_FILE);
                monopoly.getRules().copyFrom(rules);
            } catch (Exception e) {
                frame.showMessageDialog("ERROR", "Failed to load rules from " + RULES_FILE + "\n " + e.getClass().getSimpleName() + " " + e.getMessage());
            }
        }
    }


    int numImagesLoaded = 0;

    enum Images {
        monopoly_board, car, dog, iron, ship, shoe, thimble, tophat, wheelbarrow;
    }

    int [] ids = new int[Images.values().length];
    Map<Piece, Integer> pieceMap = new HashMap<>();

    @Override
    protected void init(AWTGraphics g) {
        for (int i=0; i<ids.length; i++) {
            ids[i] = g.loadImage("images/" + Images.values()[i].name() + ".png");
            numImagesLoaded++;
        }
        pieceMap.put(Piece.BOAT, Images.ship.ordinal());
        pieceMap.put(Piece.CAR, Images.car.ordinal());
        pieceMap.put(Piece.DOG, Images.dog.ordinal());
        pieceMap.put(Piece.HAT, Images.tophat.ordinal());
        pieceMap.put(Piece.IRON, Images.iron.ordinal());
        pieceMap.put(Piece.SHOE, Images.shoe.ordinal());
        pieceMap.put(Piece.THIMBLE, Images.thimble.ordinal());
        pieceMap.put(Piece.WHEELBARROW, Images.wheelbarrow.ordinal());
    }

    @Override
    protected float getInitProgress() {
        return (float)numImagesLoaded / ids.length;
    }


    @Override
    protected void paint(AWTGraphics g, int mouseX, int mouseY) {
        monopoly.paint(g, mouseX, mouseY);
    }


}

File: SwingApps/apps/cc/applets/monopoly/AWTMonopoly.kt
package cc.applets.monopoly

import cc.lib.game.AGraphics
import cc.lib.logger.LoggerFactory
import cc.lib.monopoly.Card
import cc.lib.monopoly.MoveType
import cc.lib.monopoly.Piece
import cc.lib.monopoly.Player
import cc.lib.monopoly.Player.CardChoiceType
import cc.lib.monopoly.PlayerUser
import cc.lib.monopoly.Rules
import cc.lib.monopoly.Square
import cc.lib.monopoly.Trade
import cc.lib.monopoly.UIMonopoly
import cc.lib.reflector.Reflector
import cc.lib.swing.AWTButton
import cc.lib.swing.AWTComponent
import cc.lib.swing.AWTFrame
import cc.lib.swing.AWTFrame.OnListItemChoosen
import cc.lib.swing.AWTGraphics
import cc.lib.swing.AWTLabel
import cc.lib.swing.AWTNumberPicker
import cc.lib.swing.AWTPanel
import cc.lib.swing.AWTToggleButton
import cc.lib.utils.FileUtils
import cc.lib.utils.prettify
import java.awt.BorderLayout
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.io.File
import java.util.concurrent.locks.ReentrantLock
import javax.swing.JLabel
import javax.swing.JSpinner
import javax.swing.SpinnerNumberModel
import javax.swing.event.ChangeEvent
import kotlin.concurrent.withLock
import kotlin.math.roundToInt

class AWTMonopoly internal constructor() : AWTComponent() {
	val SAVE_FILE: File
	val RULES_FILE: File
	val frame: AWTFrame
	val lock = ReentrantLock()
	val cond = lock.newCondition()

	fun showRulesPopup() {
		val rules = monopoly.rules
		val popup = AWTFrame()
		val content = AWTPanel(BorderLayout())
		val npStart = AWTNumberPicker.Builder().setLabel("Start $").setMin(500).setMax(1500).setStep(100).setValue(rules.startMoney).build { _, value ->
			rules.startMoney = value
		}
		val npWin = AWTNumberPicker.Builder().setLabel("Win $").setMin(2000).setMax(5000).setStep(500).setValue(rules.valueToWin).build(null)
		val npTaxScale = AWTNumberPicker.Builder().setLabel("Tax Scale %").setMin(0).setMax(200).setStep(25).setValue((100f * rules.taxScale).roundToInt()).build(null)
		val jailBump = AWTToggleButton("Jail Bump", rules.jailBumpEnabled)
		val jailMulti = AWTToggleButton("Jail Multiplier", rules.jailMultiplier)
		val jailMaxTurns = AWTNumberPicker.Builder().setLabel("Max Jail Turns").setMin(2).setMax(5).setValue(rules.maxTurnsInJail).build(null)
		val doublesRule = AWTToggleButton("Extra Turn on Doubles", rules.doublesRule)
		content.addTop(AWTLabel("RULES", 1, 20f, true))
		val buttons = AWTPanel(0, 1)
		val pickers = AWTPanel(1, 0)
		pickers.add(npStart)
		pickers.add(npWin)
		pickers.add(npTaxScale)
		pickers.add(jailMaxTurns)
		content.add(pickers)
		content.addRight(buttons)
		buttons.add(jailBump)
		buttons.add(jailMulti)
		buttons.add(doublesRule)
		buttons.add(object : AWTButton("Apply") {
			override fun onAction() {
				rules.valueToWin = npWin.value
				rules.taxScale = 0.01f * npTaxScale.value
				rules.startMoney = npStart.value
				rules.jailBumpEnabled = jailBump.isSelected
				rules.jailMultiplier = jailMulti.isSelected
				rules.maxTurnsInJail = jailMaxTurns.value
				rules.doublesRule = doublesRule.isSelected
				popup.closePopup()
				try {
					Reflector.serializeToFile<Any>(rules, RULES_FILE)
				} catch (e: Exception) {
					frame.showMessageDialog("ERROR", """Error save rules to file $RULES_FILE
${e.javaClass.simpleName} ${e.message}""")
				}
			}
		})
		content.addBottom(object : AWTButton("Cancel") {
			override fun onAction() {
				popup.closePopup()
			}
		})
		popup.contentPane = content
		popup.showAsPopup(frame)
	}

	init {
		setMouseEnabled(true)
		setPadding(5)
		val settings = FileUtils.getOrCreateSettingsDirectory(javaClass)
		SAVE_FILE = File(settings, "game.save")
		RULES_FILE = File(settings, "rules.save")

		frame = object : AWTFrame("Monopoly") {
			/*
            @Override
            protected void onWindowClosing() {
                if (monopoly.isGameRunning())
                    monopoly.trySaveToFile(SAVE_FILE);
            }*/
			override fun onMenuItemSelected(menu: String, subMenu: String) {
				when (menu) {
					"File" -> when (subMenu) {
						"New Game" -> showListChooserDialog(object : OnListItemChoosen {
							override fun itemChoose(item: Int) {
								showListChooserDialog(object : OnListItemChoosen {
									override fun itemChoose(pc: Int) {
										monopoly.stopGameThread()
										monopoly.initPlayers(item + 2, Piece.values()[pc])
										monopoly.newGame()
										monopoly.startGameThread()
									}

								}, "Choose Piece", *Piece.values().toList().prettify())
							}

						}, "How Many Players?", "2", "3", "4")
						"Resume"   -> if (monopoly.tryLoadFromFile(SAVE_FILE)) {
							monopoly.startGameThread()
						}
						"Rules"    -> showRulesPopup()
					}
				}
			}
		}
		frame.addMenuBarMenu("File", "New Game", "Resume", "Rules")
		frame.add(this)
		if (!frame.loadFromFile(File(settings, "monopoly.properties")))
			frame.centerToScreen(800, 600)
	}

	val monopoly: UIMonopoly = object : UIMonopoly() {
		override fun repaint() {
			this@AWTMonopoly.repaint()
		}

		override fun getImageId(p: Piece): Int {
			return pieceMap[p]!!
		}

		override val boardImageId: Int
			get() = Images.monopoly_board.boardImageId

		override fun showChooseMoveMenu(player: Player, moves: List<MoveType>): MoveType? {
			val result = arrayOfNulls<MoveType>(1)
			val strings = mutableListOf<String>()
			for (mt in moves) {
				when (mt) {
					MoveType.PURCHASE_UNBOUGHT -> {
						val sq = getPurchasePropertySquare()
						strings.add("Purchase ${sq.name.prettify()} \$${sq.price}")
					}
					MoveType.PAY_BOND -> {
						strings.add(String.format("%s $%d", mt.name.prettify(), player.jailBond))
					}
					else -> strings.add(mt.name.prettify())
				}
			}
			trySaveToFile(SAVE_FILE)
			frame.showListChooserDialog(object : OnListItemChoosen {
				override fun itemChoose(index: Int) {
					result[0] = moves[index]
					when (result[0]) {
						MoveType.PURCHASE_UNBOUGHT -> {
							showPropertyPopup("Purchase?", "Buy",
								getPurchasePropertySquare(), {
									lock.withLock { cond.signal() }
								}
							) {
								result[0] = null
								lock.withLock { cond.signal() }
							}
						}
						MoveType.PURCHASE          -> {
							showPropertyPopup("Purchase?", "Buy",
								player.square, { lock.withLock { cond.signal() } }
							) {
								result[0] = null
								lock.withLock { cond.signal() }
							}
						}
						else                       -> lock.withLock { cond.signal() }
					}
				}

				override fun cancelled() {
					if (canCancel())
						cancel()
					else
						stopGameThread()
					lock.withLock { cond.signal() }
				}
			}, "Choose Move " + player.piece, *strings.toTypedArray())
			lock.withLock { cond.await() }
			return result[0]
		}

		override fun showChooseCardMenu(player: Player, cards: List<Card>, type: CardChoiceType): Card? {
			val result = arrayOfNulls<Card>(1)
			val items = mutableListOf<String>()
			for (c in cards) {
				when (type) {
					CardChoiceType.CHOOSE_CARD_TO_MORTGAGE -> items.add(c.property.name.prettify() + " Mortgage $" + c.property.getMortgageValue(c.houses))
					CardChoiceType.CHOOSE_CARD_TO_UNMORTGAGE -> items.add(c.property.name.prettify() + " Buyback $" + c.property.mortgageBuybackPrice)
					CardChoiceType.CHOOSE_CARD_FOR_NEW_UNIT -> items.add(c.property.name.prettify() + " Unit $" + c.property.unitPrice)
				}
			}
			frame.showListChooserDialog(object : OnListItemChoosen {
				override fun itemChoose(index: Int) {
					result[0] = cards[index]
					lock.withLock { cond.signal() }
				}

				override fun cancelled() {
					if (canCancel())
						cancel()
					else
						stopGameThread()
					lock.withLock { cond.signal() }
				}
			}, player.piece.toString() + " " + type.name.prettify(), *items.toTypedArray())
			lock.withLock { cond.await() }
			return result[0]
		}

		override fun showChooseTradeMenu(player: Player, trades: List<Trade>): Trade? {
			val result = arrayOfNulls<Trade>(1)
			frame.showListChooserDialog(object : OnListItemChoosen {
				override fun itemChoose(index: Int) {
					val t = trades[index]
					val trader = t.trader
					val title = "Buy " + t.card.property.name.prettify() + " from " + trader.piece.name.prettify()
					showPropertyPopup(title, "Buy for $" + t.price, t.card.property, {
						result[0] = t
						lock.withLock { cond.signal() }
					}) { lock.withLock { cond.signal() } }
				}

				override fun cancelled() {
					if (canCancel())
						cancel()
					else
						stopGameThread()
					lock.withLock { cond.signal() }
				}
			}, "Choose Trade" + player.piece, *trades.prettify())
			lock.withLock { cond.await() }
			return result[0]
		}

		override fun showMarkSellableMenu(user: PlayerUser, sellable: List<Card>): Boolean {
			val popup = AWTFrame()
			val listener = ActionListener { e: ActionEvent? ->
				lock.withLock { cond.signal() }
				popup.closePopup()
			}
			val content = AWTPanel(BorderLayout())
			val list = AWTPanel(0, 2)
			for (card in sellable) {
				val price = user.getSellableCardCost(card)
				list.add(JLabel(card.property.name))
				val spinner = JSpinner(SpinnerNumberModel(price, 0, 2000, 50))
				spinner.addChangeListener { e: ChangeEvent? -> user.setSellableCard(card, (spinner.value as Int)) }
				list.add(spinner)
			}
			content.add(AWTLabel("Mark Sellable " + user.piece.name, 1, 20f, true), BorderLayout.NORTH)
			content.add(list, BorderLayout.CENTER)
			content.add(AWTButton("DONE", listener), BorderLayout.SOUTH)
			popup.contentPane = content
			popup.showAsPopup(frame)
			lock.withLock { cond.await() }
			return true
		}
	}

	fun showPropertyPopup(title: String, yesLabel: String, sq: Square, onBuyRunnable: Runnable, onDontBuyRunnble: Runnable): AWTFrame {
		val popup = AWTFrame()
		val content = AWTPanel(BorderLayout())
		content.add(AWTLabel(title, 1, 20f, true), BorderLayout.NORTH)
		val sqPanel: AWTComponent = object : AWTComponent() {
			override fun paint(g: AWTGraphics) {
				monopoly.drawPropertyCard(g, g.viewportWidth.toFloat(), 0f, sq!!, null)
			}
		}
		sqPanel.setPreferredSize(monopoly.DIM / 3, monopoly.DIM / 4 * 2)
		content.add(sqPanel)
		content.add(AWTPanel(1, 2, object : AWTButton(yesLabel) {
			override fun onAction() {
				popup.closePopup()
				onBuyRunnable.run()
			}
		}, object : AWTButton("Dont Buy") {
			override fun onAction() {
				popup.closePopup()
				onDontBuyRunnble.run()
			}
		}), BorderLayout.SOUTH)
		popup.contentPane = content
		popup.showAsPopup(frame)
		return popup
	}

	var numImagesLoaded = 0

	internal enum class Images {
		monopoly_board,
		car,
		dog,
		iron,
		ship,
		shoe,
		thimble,
		tophat,
		wheelbarrow;

		var boardImageId = -1
	}

	var pieceMap: MutableMap<Piece, Int> = HashMap()
	override fun init(g: AWTGraphics) {
		try {
			val rules:Rules = Reflector.deserializeFromFile(RULES_FILE)
			monopoly.rules.copyFrom(rules)
		} catch (e: Exception) {
			e.printStackTrace()
		}

		g.addSearchPath("SwingApps/images")
		for (img in Images.values()) {
			img.boardImageId = g.loadImage("${img.name}.png")
			numImagesLoaded++
		}
		pieceMap[Piece.BOAT] = Images.ship.boardImageId
		pieceMap[Piece.CAR] = Images.car.boardImageId
		pieceMap[Piece.DOG] = Images.dog.boardImageId
		pieceMap[Piece.HAT] = Images.tophat.boardImageId
		pieceMap[Piece.IRON] = Images.iron.boardImageId
		pieceMap[Piece.SHOE] = Images.shoe.boardImageId
		pieceMap[Piece.THIMBLE] = Images.thimble.boardImageId
		pieceMap[Piece.WHEELBARROW] = Images.wheelbarrow.boardImageId
	}

	override val initProgress: Float
		get() = numImagesLoaded.toFloat() / Images.values().size

	override fun paint(g: AWTGraphics) {
		monopoly.paint(g, mouseX, mouseY)
	}

	companion object {
		private val log = LoggerFactory.getLogger(AWTMonopoly::class.java)

		@JvmStatic
		fun main(args: Array<String>) {
			AGraphics.DEBUG_ENABLED = true
			//Utils.setDebugEnabled()
			AWTMonopoly()
		}
	}

}